File: src/lib/scala/Slate.Api/src/main/scala/slate/core/apis/ApiValidator.scala
/**
  * <slate_header>
  * url: www.slatekit.com
  * git: www.github.com/code-helix/slatekit
  * org: www.codehelix.co
  * author: Kishore Reddy
  * copyright: 2016 CodeHelix Solutions Inc.
  * license: refer to website and/or github
  * about: A Scala utility library, tool-kit and server backend.
  * mantra: Simplicity above all else
  * </slate_header>
  */
package slate.core.apis

import slate.common.results.{ResultSupportIn}
import slate.common.{NoResult, Todo, Result}
import slate.core.apis.support.{ApiCallCheck, ApiCallHelper, ApiCallReflect}

object ApiValidator extends ResultSupportIn {



  private def check(cmd:Request, fetcher:(Request)=> Result[(ApiCallReflect,ApiBase)])
  :(Boolean, Result[Any], Result[(ApiCallReflect,ApiBase)]) = {
    // e.g. "users.invite" = [ "users", "invite" ]
    // Check 1: at least 2 parts
    val totalParts = Option(cmd).map( c => c.parts.size).getOrElse(0)
    if( totalParts < 2) {
      (false, badRequest(Some(cmd.action + ": invalid call")), NoResult)
    }
    else {
      // Check 2: Not found ?
      val check = fetcher(cmd)
      val result = if(check.success) success(true) else failure(msg = check.msg)
      (check.success, result, check)
    }
  }

  /**
    * whether or not the api call represented by the area.api.action exists. e.g. "app.users.invite"
    * and the parameters are valid.
    *
    * @param cmd       : the command input
    * @return
    */
  def validateCall(cmd:Request,
                   fetcher:(Request)=> Result[(ApiCallReflect,ApiBase)],
                   allowSingleDefaultParam:Boolean = false): Result[ApiCallCheck] =
  {
    val fullName  = cmd.fullName
    val apiArea   = cmd.area
    val apiName   = cmd.name
    val apiAction = cmd.action
    val args      = cmd.args.get
    val checkResult = check(cmd, fetcher)

    if(!checkResult._1){
      badRequest[ApiCallCheck](msg =Some("bad request : " + fullName + ": inputs not supplied"))
    }
    else {
      val result = checkResult._3.get
      val callReflect = result._1
      val api = result._2

      //if(!callReflect.hasArgs && args.size() > 0)
      //  return badRequest("bad request : " + fullName + ": takes 0 inputs")
      // Check 3: 1 param with default argument.
      if (allowSingleDefaultParam && callReflect.isSingleDefaultedArg() && args.size == 0) {
        success(data = new ApiCallCheck(true, apiArea, apiName, apiAction, false, api, cmd))
      }
      // 4a: Param: Raw ApiCmd itself!
      else if (callReflect.isSingleArg() && (callReflect.paramList(0).typeName == "ApiCmd"
        || callReflect.paramList(0).typeName == "Request")) {
        success(data = new ApiCallCheck(true, apiArea, apiName, apiAction, false, api, cmd))
      }
      // 4b: Params - check args needed
      else if (!allowSingleDefaultParam && callReflect.hasArgs && args.size == 0)
        badRequest[ApiCallCheck](msg =Some("bad request : " + fullName + ": inputs not supplied"))

      // 4c: Params - ensure matching args
      else if (callReflect.hasArgs) {
        val argCheck = ApiCallHelper.validateArgs(callReflect, args)
        if(argCheck.success) {
          success(data = new ApiCallCheck(true, apiArea, apiName, apiAction, false, api, cmd))
        }
        else
          badRequest[ApiCallCheck](msg =Some("bad request : " + fullName + ": inputs not supplied"))
      }
      else
        success(data = new ApiCallCheck(true, apiArea, apiName, apiAction, false, api, cmd))
    }
  }
}

File: src/lib/kotlin/slatekit-apis/src/main/kotlin/slatekit/apis/helpers/ApiValidator.kt
/**
 * <slate_header>
 * url: www.slatekit.com
 * git: www.github.com/code-helix/slatekit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 * about: A tool-kit, utility library and server-backend
 * mantra: Simplicity above all else
 * </slate_header>
 */

package slatekit.apis.helpers

import slatekit.apis.core.Target
import slatekit.apis.core.Action
import slatekit.common.*
import slatekit.common.requests.Request
import slatekit.results.Failure
import slatekit.results.Notice
import slatekit.results.Success
import slatekit.results.builders.Notices

object ApiValidator {

    /**
     * Checks the "route" ( area.api.action ) is valid.
     */
    fun check(req: Request, fetcher: (Request) -> Notice<Target>): Notice<Target> {
        // e.g. "users.invite" = [ "users", "invite" ]
        // Check 1: at least 2 parts
        val totalParts = req.parts.size
        return if (totalParts < 2) {
           Notices.invalid(req.action + ": invalid call")
        } else {
            // Check 2: Not found ?
            val check = fetcher(req)
            check
        }
    }

    /**
     * whether or not the api call represented by the area.api.action exists. e.g. "app.users.invite"
     * and the parameters are valid.
     *
     * @param req : the command input
     * @return
     */
    fun validateCall(
            req: Request,
            fetcher: (Request) -> Notice<Target>,
            allowSingleDefaultParam: Boolean = false
    ): Notice<Target> {
        val fullName = req.fullName
        val args = req.data
        val apiRefCheck = check(req, fetcher)

        return when (apiRefCheck) {
            is Failure -> Notices.invalid("bad request : $fullName: inputs not supplied")
            is Success -> {
                val apiRef = apiRefCheck.value
                val action = apiRef.action

                // 1 param with default argument.
                if (allowSingleDefaultParam && action.isSingleDefaultedArg() && args.size() == 0) {
                    Success(apiRef)
                }
                // Param: Raw ApiCmd itself!
                else if (action.isSingleArg() && action.paramsUser.isEmpty()) {
                    Success(apiRef)
                }
                // Params - check args needed
                else if (!allowSingleDefaultParam && action.hasArgs && args.size() == 0)
                    Notices.invalid<Target>("bad request : $fullName: inputs not supplied")

                // Params - ensure matching args
                else if (action.hasArgs) {
                    val argCheck = validateArgs(action, args)
                    if (argCheck.success) {
                        Success(apiRef)
                    } else
                        Notices.invalid<Target>("bad request : $fullName: inputs not supplied")
                } else
                    Success(apiRef)
            }
        }
    }

    private fun validateArgs(action: Action, args: Inputs): Notice<Boolean> {
        var error = ": inputs missing or invalid "
        var totalErrors = 0

        // Check each parameter to api call
        for (index in 0 until action.paramsUser.size) {
            val input = action.paramsUser[index]
            // parameter not supplied ?
            val paramName = input.name!!
            if (!args.containsKey(paramName)) {
                val separator = if (totalErrors == 0) "( " else ","
                error += separator + paramName
                totalErrors += 1
            }
        }
        // Any errors ?
        return if (totalErrors > 0) {
            error = "$error )"
            Notices.invalid("bad request: action " + action.name + error)
        } else {
            Success(true)
        }
    }
}
